src/java/org/apache/cassandra/service/consensus/migration/TableMigrationState.java

```diff
@@ -0,0 +1,360 @@
+import static org.apache.cassandra.utils.CollectionSerializers.serializedCollectionSize;
+import static org.apache.cassandra.utils.CollectionSerializers.serializedMapSize;
+
+// TODO Move this into the schema for the table once this is based off of TrM
+public class TableMigrationState
+{
+    private static final Logger logger = LoggerFactory.getLogger(TableMigrationState.class);
+
+    @Nonnull
+    public final String keyspaceName;
+
+    @Nonnull
+    public final String tableName;
+
+    @Nonnull
+    public final TableId tableId;
+
+    @Nonnull
+    public final ConsensusMigrationTarget targetProtocol;
+
+    @Nonnull
+    public final List<Range<Token>> migratedRanges;
+
+    /*
+     * Necessary to track which ranges started migrating at which epoch
+     * in order to know whether a repair qualifies in terms of finishing
+     * migration of the range.
+     */
+    @Nonnull
+    public final NavigableMap<Epoch, List<Range<Token>>> migratingRangesByEpoch;
+
+    @Nonnull
+    public final List<Range<Token>> migratingRanges;
+
+    @Nonnull
+    public final List<Range<Token>> migratingAndMigratedRanges;
+
+    public TableMigrationState(@Nonnull String keyspaceName,
+                               @Nonnull String tableName,
+                               @Nonnull TableId tableId,
+                               @Nonnull ConsensusMigrationTarget targetProtocol,
+                               @Nonnull Collection<Range<Token>> migratedRanges,
+                               @Nonnull Map<Epoch, List<Range<Token>>> migratingRangesByEpoch)
+    {
+        this.keyspaceName = keyspaceName;
+        this.tableName = tableName;
+        this.tableId = tableId;
+        this.targetProtocol = targetProtocol;
+        this.migratedRanges = ImmutableList.copyOf(normalize(migratedRanges));
+        this.migratingRangesByEpoch = ImmutableSortedMap.copyOf(
+        migratingRangesByEpoch.entrySet()
+                              .stream()
+                              .map(entry -> new AbstractMap.SimpleEntry<>(entry.getKey(), ImmutableList.copyOf(normalize(entry.getValue()))))
+                              .collect(Collectors.toList()));
+        this.migratingRanges = ImmutableList.copyOf(normalize(migratingRangesByEpoch.values().stream().flatMap(Collection::stream).collect(Collectors.toList())));
+        this.migratingAndMigratedRanges = ImmutableList.copyOf(normalize(ImmutableList.<Range<Token>>builder().addAll(migratedRanges).addAll(migratingRanges).build()));
+    }
+
+    public TableMigrationState reverseMigration(ConsensusMigrationTarget target, Epoch epoch)
+    {
+        IPartitioner partitioner = DatabaseDescriptor.getPartitioner();
+        Range<Token> fullRange = new Range<>(partitioner.getMinimumToken(), partitioner.getMinimumToken());
+        List<Range<Token>> allTouched = new ArrayList<>(migratedRanges);
+        allTouched.addAll(migratingRanges);
+        allTouched = Range.deoverlap(allTouched);
```

Comment:
As far as I can tell this still has the issue that you can reverse migration for partially migrated ranges. For those ranges some keys are on the old system and some are on the new one so it's not a safe operation to reverse direction.

Reply:
Actually I think this is safe/fine since all keys already migrated will be key migrated before being used. In fact you could reverse several times and it should be fine? You will migrate on one system before using the other and you are always going to see the latest value for a key because either you picked the correct system, or you did the key migration first.